### ISSUE TITLE
Magit hangs on staging/discarding hunks on remote repository from windows machine.

### ISSUE THREAD

**Issue by rkit1** [magit/magit#3624]
Magit 2.13.0 on Emacs 26.1 on Windows 10 using plink as ssh client. Remote repo tested on Ubuntu server and Debian.

1. Open magit-diff, select chunk, press s or k.
2. Status line shows `Tramp: Opening connection for <host> using plink...done`
3. Nothing happens for minutes.

After pressing C-g, I am able to see magit-process buffer. It shows 
`run git … "apply" "--cached" "-p0" "--ignore-space-change" "-"`

`ps aux` on server shows this process running

In debugger I found out that execution is stuck in endless loop on those lines in magit-run-git-with-input
```
        (while (and magit-this-process
                    (eq (process-status magit-this-process) 'run))
(sleep-for 0.005)))
```
As far as I can tell, magit does not close stdin of git process, so in never exits.

**Comment by tarsius** [comment]
> As far as I can tell, magit does not close stdin of git process, so in never exits.

I don't think that's the issue. We used the same approach (until e62aff3fb3e207ade24ebbc879d748bf6a2ed708) that now is being used just over tramp when applying locally too, and would have noticed if that didn't work.

**Comment by ferfebles** [comment]
Using plink too

**Comment by ferfebles** [comment]
Magit launches a git process in the remote host that never ends. This is the process info:

```
# ps -fea | grep git
myuser     19096 19093  0 12:39 pts/2    00:00:00 git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false -c i18n.logOutputEncoding=UTF-8 apply --cached -p0 --ignore-space-change -
```
The git process has been running for several minutes.

**Comment by tarsius** [comment]
Please try if it works with another ssh client.

**Comment by ferfebles** [comment]
Tested with pscp (from putty installer). Same problem

**Comment by tarsius** [comment]
Try `(setq magit-process-connection-type t)`.

**Comment by rkit1** [comment]
> 
> 
> Try `(setq magit-process-connection-type t)`.

Didn't help

**Comment by ferfebles** [comment]
Tried with (setq magit-process-connection-type t), but hangs again.

**Comment by tarsius** [comment]
> As far as I can tell, magit does not close stdin of git process, so in never exits.

Actually it does by calling `process-send-eof` in `magit-start-process` (which is called by `magit-start-git`).

**Comment by gbrunick** [comment]
Unfortunately, it appears that  `process-send-eof` not working on Windows is know issue with TRAMP.  See comments related to `tramp-test29-start-file-process` at the end of:  [tramp-tests.el](http://git.savannah.gnu.org/cgit/emacs.git/plain/test/lisp/net/tramp-tests.el?h=emacs-26).

**Comment by prashantvithani** [comment]
I just noticed this happening in Linux as well. In addition, the operations like `stage`, `unstage` are also hung. 

Git: `2.25.1`,
Remote OS: `Ubuntu 20.04.2`
Host OS: `"Arch Linux" Linux 5.12.12-arch1-1 x86_64`
Emacs: `28.0.50 master b16b4d730`
Magit: b68a760c

**Comment by prashantvithani** [comment]
It seems to have been stuck [here][https://github.com/magit/magit/blob/f592e367d7602106d6f8f8349148bd31a94ba231/lisp/magit-process.el#L454-L455] though

**Comment by pabl0** [comment]
I can confirm that this issue also happens with `ssh.exe` (Microsoft's OpenSSH port).

I believe the relevant Emacs bugs are: [43226](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=43226) and [6771](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=6771).

Isn't `process-send-eof` a core Emacs function (implemented in C) not part of Tramp?

`magit-process-connection-type`  seems to be `t` by default. Setting it to `nil` does not seem to make a difference. Besides, are PTYs even supported on Windows Emacs yet?

Instead of a FAQ entry in the manual,would it be possible to disable the support for staging individual hunks and emitting some sort of message, until this bug has been resolved in Emacs? This hang is quite surprising, since most other functionality seem to work just fine, including staging whole files (since it just simply does `git add`).

One workaround could be to send a temporary file to be fed to git apply, but that feels rather ugly.

**Comment by pabl0** [comment]
I have a hunch that this could have something to do with cmdproxy.exe in Windows.

If I do a simple test like:

```
(let ((start-file-process "eoftest" (get-buffer-create "*eoftest*")
			  "ssh" "<hostname>" "cat"))
  (sit-for 10)
  (process-send-string sshz "foo")
  (sit-for 5)
  (process-send-eof sshz))
```
and monitor in the remote host, I can trace the process and see "foo" coming as input, and the process is gone after EOF.

However if I attempt the following:
```
(let ((default-directory "/sshx:<hostname>:/home/my/git/dir")
      (sshz start-file-process "eoftest" (get-buffer-create "*eoftest*")
			  "cat"))
  (sit-for 10)
  (process-send-string sshz "foo")
  (sit-for 5)
  (process-send-eof sshz))
```
tramp will start the process with cmdproxy.exe and cat get's stuck forever – it seems it does not even receive "foo" as input.

**Comment by rkit1** [comment]
> tramp will start the process with cmdproxy.exe and cat get's stuck forever – it seems it does not even receive "foo" as input.

Line buffering issue?

**Comment by pabl0** [comment]
Did some debugging with edebug to verify that everything else except `process-send-eof `seem to work fine: ran `strace -p PID`for the git process on the remote server while on breakpoint and saw that the process received the patch from `process-send-region`. However, `process-send-eof` did not cause EOF to be received, as expected.

Then I inserted `(process-send-string process "\004")` before line 594, and after that adding a hunk from the magit interface worked just fine! Perhaps the fix is simply to replace `process-send-eof` with simply sending EOF as a string?

https://github.com/magit/magit/blob/f592e367d7602106d6f8f8349148bd31a94ba231/lisp/magit-process.el#L591-L594

**Comment by perezandrew477** [comment]
Using pabl0's suggestion of sending EOF as a string `(process-send-string process "\004")` fixed the issue for me. I can now stage/discard hunks without an issue from Windows on remote hosts.

**Comment by tarsius** [comment]
I've applied a kludge along these lines.

**Comment by schellj** [comment]
FWIW, I'm having this issue over tramp to a CentOS 7 VM. Neither the change in https://github.com/magit/magit/commit/2286a6397485f815de70e8105eda90fd0caed4e7 (i.e., `(process-send-string process "")`), nor `(process-send-string process "\004")`, nor `(process-send-eof process)` fixes the issue for me. I'm not sure what else might work, but I'm happy to try out suggestions.

**Comment by tarsius** [comment]
The last fix for this issue is mentioned at https://github.com/magit/magit/issues/4720#issuecomment-1874069488. Are you're Tramp and Magit versions recent enough?

**Comment by schellj** [comment]
@tarsius I believe so; I have Tramp 2.6.2.29.2 and Magit 20240125.1835.

**Comment by KevinBrowne** [comment]
@schellj That's the built-in version of Tramp that shipped with Emacs; the last two numbers, `29.2`, are your Emacs version. Despite what the version number might suggest, it does not include the change that the latest Magit fix depends on, as I discovered when I hit the same problem.

You should install the Tramp package, version 2.6.2 or higher (it's on GNU ELPA), and ensure it's loaded in preference to the built-in (`M-x tramp-version` should report the package's version, not the built-in version).

**Comment by schellj** [comment]
Installing the latest version from ELPA (`2.9.2.1`) fixed the issue. Thanks @KevinBrowne!

**Comment by marc-h38** [comment]
> Installing the latest version from ELPA (2.9.2.1) fixed the issue. 

There is no 2.9.2.1 version, pretty sure you meant 2.6.2.1.

I'm using macOS and I have the opposite experience: a apparent regression with an even newer tramp!

This issue (or something very similar to it) just appeared in Emacs 29.2 which ships with tramp version 2.6.2.29.2

I still have Emacs 29.1  (with tramp-version 2.6.0.29.1) installed on the same mac and it still works just fine.

I cannot stage or unstage hunks either: the same, remote `git apply` command hangs forever. Staging entire files is never an issue in any configuration.

I just upgraded to the latest magit version 540c744aed13, no difference.

I temporarily hacked magit-process.el to enable the 2286a639748 , `^D` kludge on my mac too and that made no difference either.

I upgraded tramp in Emacs 29.1 to version 2.6.2.1 using ELPA and everything still works fine.
So maybe this is a (different?) regression between tramp 2.6.2.1 and 2.6.2.29.2?

I tried downgrading tramp in Emacs 29.2 but the higher tramp version available seems to always takes precedence and macOS System Integrity Protection does not let me change files in /Applications/, not even as root. Any trick to overcome this?

I'm comfortable git bisecting. Is it hard to run tramp from source? With Emacs 29.1 since 29.2 does not let me...

I'm going to stick to Emacs 29.1 for the moment, already spent a bit too many hours investigating this.

Very funny how this 2018 issue seems to just resurrect left and right.

**Comment by tarsius** [comment]
@marc-h38 As mentioned [above](https://github.com/magit/magit/issues/3624#issuecomment-1930440006) the fix (https://github.com/emacsmirror/emacs/commit/2a1a7a8524c0307c09c91e89816b2b2b8bfb85bc) is not included in the version distributed with Emacs 29.2, despite what the version strings suggests.

Only "critical bugfixes" and "changes that come with absolutely no risk" were allowed on the `emacs-29` branch, while work on Emacs 29.2 was ongoing and the Tramp fix was created. It appears that it was decided that this bugfix does not fall into either of these categories.

So users of Emacs 29.2 have no choice but to install Tramp from GNU ELPA. You have said that you have already done that and that it didn't work. So now you have to find out why it doesn't work for you (but does for others as we have seen above).

One possibility is that you are not actually using the Tramp version that you think you are using. Maybe something in Emacs loads the built-in version of Tramp, *before* the packages installed by `package.el` are put on the load path. Or you, or *something*, has somehow messed up the order of the `load-path`.

Try:
1. Is `tramp-pipe-stty-settings` defined?
2. Does `(find-library-name "tramp-sh")` return a library at the expected path? Does it contain the definition of `tramp-pipe-stty-settings`?
3. Does `(find-library--from-load-history "tramp-sh")` return a library at the expected path? Does it contain the definition of `tramp-pipe-stty-settings`?

Also try setting `magit-tramp-pipe-stty-settings` to `"-icanon min 0 time 1"` instead of `""`.

Note that while this fix is absent from Emacs 29.2, it has landed on the `emacs-29` branch after that release was cut (in https://github.com/emacsmirror/emacs/commit/115908469d30f8c40689673312f72b44c1631c6b). So you could build your Emacs from that branch. It's currently in a "non-critical bugfixes that might be a bit risk are allowed too" phase.

**Comment by marc-h38** [comment]
First, thanks for the detailed and insightful explanation. I will try to find some time in the next few days to test more and raise up to your level of commitment.

> One possibility is that you are not actually using the Tramp version that you think you are using. Maybe something in Emacs loads the built-in version of Tramp, before the packages installed by package.el are put on the load path. Or you, or something, has somehow messed up the order of the load-path.

That's definitely been one of my concerns. Considering my unfortunate inability (or ignorance) to disable built-in packages, I've always been worried about what happens when multiple versions are available at the same time. What I've done so far to tell: `Ctrl-h, f, tramp-somefunction` and then click on the link to the .el source file that shows up. Can this be trusted as a way to tell what exactly what I'm running? Hopefully for the entire package and not just for one .el file?

More generally speaking, is it "first come first serve" or more like "highest version wins"? If you (or anyone else) knows or can quickly find a good "Emacs packages for noobies" guide to recommend then by all means please share. If the best there is is just  `Ctrl-h, f, find-library-name` or some other built-in documentation then so be it - but then please share anyway to save me the time to look for a more beginner-oriented one. Thanks in advance!

**Comment by tarsius** [comment]
> `Ctrl-h, f, tramp-somefunction` and then click on the link to the .el source file that shows up. Can this be trusted as a way to tell what exactly what I'm running?

I think so (but `find-library--from-load-history` would be more trustworthy). Ultimately this uses `symbol-file`, which uses `load-history` (not `load-path`). So this should reflect what actually happened in the past (not what would happen now). And that's what we need when we want to determine what file some symbol has actually been loaded from.

> Hopefully for the entire package and not just for one .el file?

If the package consists of multiple libraries, then a franken version may be in effect with some libraries loaded from one version of the package, and other libraries from another version.

You can only trust it to provide the file `tramp-somefunction` was loaded from. And since loading a file loads all the things in that file, you can also be *fairly* sure that does were loaded from that file too. (Of course another file may later be loaded, which re-defines some of these things.)

> More generally speaking, is it "first come first serve" or more like "highest version wins"?

"First on the `load-path`, at the time the request is made."

> If you (or anyone else) knows or can quickly find a good "Emacs packages for noobies" guide to recommend then by all means please share. If the best there is is just `Ctrl-h, f, find-library-name` or some other built-in documentation then so be it - but then please share anyway to save me the time to look for a more beginner-oriented one. Thanks in advance!

I am not aware of such documentation, and if it existed, it would be rather complicated (or else full of over-simplifications, which may or may not still be useful).

**Comment by marc-h38** [comment]
When I start  `/Applications/Emacs\ 29.1-1.app/Contents/MacOS/Emacs`, `load-path` includes `~/.emacs.d/elpa/tramp-2.6.2.1` as shown below.

But when I start  `/Applications/Emacs\ 29.2.app/Contents/MacOS/Emacs` instead, `load-path` is basically the same, except it's MISSING `~/.emacs.d/elpa/tramp-2.6.2.1`! All the other `/elpa/` directories are still there.

Could this be because `/Applications/Emacs 29.2.app/Contents/Resources/lisp/net/trampver.el.gz` has the wrong content `(defconst tramp-version "2.6.2.29.2"...`, where `2.6.2.29.2 >= 2.6.2.1` numerically when the code 2.6.2.29.2 is actually **older**? So the packaging code compares these wrong versions and skips adding the newer 2.6.2.1 elpa in the load-path, wrongly believing it's older than the built-in?

Any relevant trace or debug clue? BTW I found the macOS security setting that lets me modify built-in files.

```
...
"/Users/me/.emacs.d/elpa/magit-20240211.1712"
"/Users/me/.emacs.d/elpa/git-commit-20240123.1358"
"/Users/me/.emacs.d/elpa/magit-section-20240211.1319"
"/Users/me/.emacs.d/elpa/dash-20240103.1301"
"/Users/me/.emacs.d/elpa/tramp-2.6.2.1"            <========== Missing when start Emacs 29.2
"/Users/me/.emacs.d/elpa/transient-20240201.1100"
"/Users/me/.emacs.d/elpa/with-editor-20240101.2226"
"/Users/me/.emacs.d/elpa/compat-29.1.4.4"
"/Library/Application Support/Emacs/29.1/site-lisp"
"/Library/Application Support/Emacs/site-lisp"
"/Applications/Emacs 29.1-1.app/Contents/Resources/lisp"
...
```

**Comment by tarsius** [comment]
> Could this be because `/Applications/Emacs 29.2.app/Contents/Resources/lisp/net/trampver.el.gz` has the wrong content `(defconst tramp-version "2.6.2.29.2"...`, where `2.6.2.29.2 >= 2.6.2.1` numerically when the code 2.6.2.29.2 is actually **older**? So the packaging code compares these wrong versions and skips adding the newer 2.6.2.1 elpa in the load-path, wrongly believing it's older than the built-in?

Good job digging that up! That seems very likely. I had a quick look at `package.el` but couldn't make out where that would be happening -- but nevertheless this seems very likely to me.

An obvious workaround would be to just manually add that to the `load-path` yourself. Instead of just editing the variable directly, you should do that by loading the `tramp-autoloads.el` file within that directory.

I think it was a mistake to not include this bugfix in the version of Tramp that is included in Emacs 29.2. Now it seems we have discovered that it additionally isn't possible to upgrade to a version that does contain the required bugfix. If our conclusion above is correct, I think the only way to *fix* that is to create a new Tramp release, which is larger than 2.6.2.29.2. Could you please suggest this to Tramp's maintainer?

**Comment by schellj** [comment]
If you use `use-package` for your config, what worked for me was to put a `use-package tramp` with a hardcoded `load-path` call before my `use-package emacs`. Obviously, this is a bit brittle as it won't work once the tramp version changes again. There's probably a more robust solutions.

e.g.,
```
(use-package tramp
  :load-path "~/.emacs.d/elpa/tramp-2.6.2.1"
...
)

(use-package emacs
...
)
```

**Comment by marc-h38** [comment]
[ answering different people at once ]

> `(use-package tramp  :load-path "~/.emacs.d/elpa/tramp-2.6.2.1")`

Thanks, this simple one does the trick! I think I will stick to this workaround for now.


> Obviously, this is a bit brittle as it won't work once the tramp version changes again.

In that case the non-existing load-path parameter will be silently ignored (tested). Assuming our `older 2.6.2.29.2 >= newer 2.6.2.1` package theory is right, at that point there will be just some hopefully harmless "garbage" to collect in `init.el` later.


> I think the only way to fix that is to create a new Tramp release, which is larger than 2.6.2.29.2. Could you please suggest this to Tramp's maintainer?

cc: @albinus, hopefully Github notifications don't get routed to /dev/null :-)

----------

So, `use-package`: yet another way to load code, fascinating...

Another packaging fact I found very confusing: even when you comment out `require` `use-package` etc, Ctrl-h f and others still find all the code you're trying NOT to load. This makes troubleshooting much harder because it feels like you have have zero control on what's happening and when.

> I had a quick look at package.el but couldn't make out where that would be happening 

If even _you_ can't, imagine how it is for me and other "plain" users like me...

To be fair package management is really hard: https://medium.com/@sdboyer/so-you-want-to-write-a-package-manager-4ae9c17d9527
Yet I feel like Emacs (and tramp?) could try a bit "harder" here. For a start it should be simplified; probably stop trying to be too smart. Less opaque "convenience"; simple and  clear error messages instead.

> I am not aware of such documentation, and if it existed, it would be rather complicated (or else full of over-simplifications, which may or may not still be useful).

Unfortunately. I can see this now.

**Comment by albinus** [comment]
Marc ***@***.***> writes:

Hi,

>     I think the only way to fix that is to create a new Tramp release,
>     which is larger than 2.6.2.29.2. Could you please suggest this to
>     Tramp's maintainer?

Roughly planned, this will happen end of June. Until then, there will be
a monthly new Tramp 2.6.2 ELPA release.

Emacs' 29 version of package.el doesn't support such problems
out-of-the-box. Hopefully, Emacs 30 will be better with this.

Or perhaps use-package can be extended appropriately. What I have in
mind is to extend the :pin tag that the chosen package archive has
preference also over builtin packages. That is, handle "builtin" like
the other package archives, "gnu", nongnu" etc.

> Yet I feel like Emacs (and tramp?) could try a bit "harder" here. For
> a start it should be simplified; probably stop trying to be too smart.
> Less opaque "convenience"; simple and clear error messages instead.

Hmm. I don't see how Tramp could contribute better to the package.el
machinery.

Best regards, Michael.

**Comment by albinus** [comment]
Philipp Middendorf ***@***.***> writes:

Hi Philipp,

>     Note that while this fix is absent from Emacs 29.2, it has landed
>     on the emacs-29 branch after that release was cut (in
>     ***@***.***). So you could build your Emacs from
>     that branch. It's currently in a "non-critical bugfixes that might
>     be a bit risk are allowed too" phase.
>
> I have applied this patch to 29.2, but the problem persists,
> unfortunately (tramp-version does say 2.6.3-pre).

tramp-version 2.6.3-pre isn't precise enough, it is a floating version
during development. So we don't know whether the Tramp patch has been
applied.

You'll be better to install the recent Tramp version (2.6.2.2) on top of
Emacs 29.2.

Note, that this problem requires also a magit patch. The recent magit
20240320.2336 on MELPA seems to contain this.

Best regards, Michael.

**Comment by pabl0** [comment]
Emacs 29.3 was just released as a emergency security fix release. It includes Tramp with this confusing version number 2.6.3-pre, which sadly does not seem to be recent enough to fix this annoying issue.

**Comment by albinus** [comment]
Henrik Ahlgren ***@***.***> writes:

Hi,

> Emacs 29.3 was just released as a emergency security fix release. It
> includes Tramp with this confusing version number 2.6.3-pre, which
> sadly does not seem to be recent enough to fix this annoying issue.

Tramp 2.6.3-pre is an annoying version number, indeed. It happened due
to miscommunication during the Emacs release process.

I can assure you, that the magit relevant Tramp fix is contained in
Emacs 29.3.

Best regards, Michael.

**Comment by pabl0** [comment]
Looks like you are correct. Apparently upgrading Magit to 20240320.2336 fixes this even with Tramp shipped with Emacs 29.3, but the problem is only fixed after Emacs restart. Installing tramp-2.6.2.2 from ELPA does not make much difference, since the built-in version is used by default.

**Comment by albinus** [comment]
Henrik Ahlgren ***@***.***> writes:

Hi Henrik,

> Looks like you are correct. Apparently upgrading Magit to
> 20240320.2336 fixes this even with Tramp shipped with Emacs 29.3, but
> the problem is only fixed after Emacs restart. Installing
> tramp-2.6.2.2 from ELPA does not make much difference, since the
> built-in version is used by default.

I'm not so fluent with ELPA use. Perhaps there is a way to configure
your Emacs to use always the ELPA provided package.

Best regards, Michael.
